File: src/client/java/io/github/romeoahmed/cursedoath/client/animation/CastingAnimation.java
package io.github.romeoahmed.cursedoath.client.animation;

import com.zigythebird.playeranim.animation.PlayerAnimationController;
import com.zigythebird.playeranim.api.PlayerAnimationAccess;
import com.zigythebird.playeranim.api.PlayerAnimationFactory;
import com.zigythebird.playeranimcore.api.firstPerson.FirstPersonMode;
import com.zigythebird.playeranimcore.enums.PlayState;
import io.github.romeoahmed.cursedoath.CursedOath;
import io.github.romeoahmed.cursedoath.technique.Technique;
import net.minecraft.resources.Identifier;
import net.minecraft.world.entity.Avatar;
import org.jspecify.annotations.Nullable;

/// PAL playback follows confirmed cast stages; it does not drive simulation.
public final class CastingAnimation {
    private static final int LAYER_PRIORITY = 1500;
    private static final Identifier LAYER = CursedOath.id("casting");

    private CastingAnimation() {}

    public static void initialize() {
        PlayerAnimationFactory.ANIMATION_DATA_FACTORY.registerFactory(LAYER, LAYER_PRIORITY, player -> {
            var controller = new PlayerAnimationController(player, (state, animation, time) -> PlayState.STOP);
            controller.setFirstPersonMode(FirstPersonMode.THIRD_PERSON_MODEL);
            return controller;
        });
    }

    public static void start(Avatar player, Technique technique, boolean release, float elapsedTicks) {
        String name = switch (technique) {
            case BLUE -> "attract";
            case RED -> "repel";
            case PURPLE -> "fusion";
            case DISMANTLE -> "dismantle";
            case CLEAVE -> "cleave";
            case HEAL -> "seal";
            case UNLIMITED_VOID -> "void";
            case MALEVOLENT_SHRINE -> "shrine";
            case INFINITY, SIMPLE_DOMAIN, AMPLIFICATION -> null;
        };
        if (name == null) return;
        var controller = controller(player);
        if (controller != null)
            controller.triggerAnimation(CursedOath.id(release ? name + "_release" : name), elapsedTicks);
    }

    public static void stop(Avatar player) {
        var controller = controller(player);
        if (controller != null) controller.stopTriggeredAnimation();
    }

    private static @Nullable PlayerAnimationController controller(Avatar player) {
        var layer = PlayerAnimationAccess.getPlayerAnimationLayer(player, LAYER);
        return layer instanceof PlayerAnimationController controller ? controller : null;
    }
}

File: src/client/kotlin/io/github/romeoahmed/cursedoath/client/animation/CastingAnimation.kt
package io.github.romeoahmed.cursedoath.client.animation

import com.zigythebird.playeranim.animation.PlayerAnimationController
import com.zigythebird.playeranim.api.PlayerAnimationAccess
import com.zigythebird.playeranim.api.PlayerAnimationFactory
import com.zigythebird.playeranimcore.api.firstPerson.FirstPersonMode
import com.zigythebird.playeranimcore.enums.PlayState
import io.github.romeoahmed.cursedoath.CursedOath
import io.github.romeoahmed.cursedoath.technique.Technique
import net.minecraft.world.entity.Avatar

/** Client-only PAL adapter; animation does not drive damage or movement. */
object CastingAnimation {
    private const val LAYER_PRIORITY = 1500
    private val layer = CursedOath.id("casting")

    fun initialize() {
        PlayerAnimationFactory.ANIMATION_DATA_FACTORY.registerFactory(layer, LAYER_PRIORITY) { player ->
            PlayerAnimationController(player) { _, _, _ -> PlayState.STOP }.apply {
                setFirstPersonMode(FirstPersonMode.THIRD_PERSON_MODEL)
            }
        }
    }

    fun start(
        player: Avatar,
        technique: Technique,
        release: Boolean,
        elapsedTicks: Float,
    ) {
        val name =
            when (technique) {
                Technique.BLUE -> "attract"
                Technique.RED -> "repel"
                Technique.PURPLE -> "fusion"
                Technique.DISMANTLE -> "dismantle"
                Technique.CLEAVE -> "cleave"
                Technique.HEAL -> "seal"
                Technique.INFINITY -> return
            }
        controller(player)?.triggerAnimation(CursedOath.id(if (release) "${name}_release" else name), elapsedTicks)
    }

    fun stop(player: Avatar) {
        controller(player)?.stopTriggeredAnimation()
    }

    private fun controller(player: Avatar) =
        PlayerAnimationAccess.getPlayerAnimationLayer(player, layer) as? PlayerAnimationController
}
